server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentKeyUpdaterDao.java

```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  public Map<String, String> checkModuleKeysBeforeRenaming(String projectUuid, String stringToReplace, String replacementString) {
+    SqlSession session = mybatis.openSession(false);
+    ComponentKeyUpdaterMapper mapper = session.getMapper(ComponentKeyUpdaterMapper.class);
+    Map<String, String> result = Maps.newHashMap();
+    try {
+      Set<ResourceDto> modules = collectAllModules(projectUuid, stringToReplace, mapper);
+      for (ResourceDto module : modules) {
+        String newKey = computeNewKey(module.getKey(), stringToReplace, replacementString);
+        if (mapper.countResourceByKey(newKey) > 0) {
+          result.put(module.getKey(), "#duplicate_key#");
+        } else {
+          result.put(module.getKey(), newKey);
+        }
+      }
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+    return result;
+  }
+
+  public static void checkIsProjectOrModule(ComponentDto component) {
+    checkArgument(PROJECT_OR_MODULE_QUALIFIERS.contains(component.qualifier()), "Component updated must be a module or a key");
+  }
+
+  /**
+   *
+   * @return a map with currentKey/newKey is a bulk update was executed
+   */
+  public Map<String, String> simulateBulkUpdateKey(DbSession dbSession, String projectUuid, String stringToReplace, String replacementString) {
+    return collectAllModules(projectUuid, stringToReplace, mapper(dbSession))
+      .stream()
+      .collect(Collectors.toMap(
+        ResourceDto::getKey,
+        component -> {
+          String newKey = computeNewKey(component.getKey(), stringToReplace, replacementString);
+          checkModuleKey(newKey);
+          return newKey;
+        }));
+  }
+
+  /**
+   * @return a map with the component key as key, and boolean as true if key already exists in db
+   */
+  public Map<String, Boolean> checkComponentKeys(DbSession dbSession, List<String> newComponentKeys) {
+    return newComponentKeys.stream().collect(Collectors.toMap(Function.identity(), key -> mapper(dbSession).countResourceByKey(key) > 0));
+  }
+
+  public static String computeNewKey(String key, String stringToReplace, String replacementString) {
+    return key.replace(stringToReplace, replacementString);
+  }
+
+  public void bulkUpdateKey(DbSession session, String projectUuid, String stringToReplace, String replacementString) {
+    ComponentKeyUpdaterMapper mapper = session.getMapper(ComponentKeyUpdaterMapper.class);
+    // must SELECT first everything
+    Set<ResourceDto> modules = collectAllModules(projectUuid, stringToReplace, mapper);
+    checkNewNameOfAllModules(modules, stringToReplace, replacementString, mapper);
+    Map<ResourceDto, List<ResourceDto>> allResourcesByModuleMap = Maps.newHashMap();
+    for (ResourceDto module : modules) {
+      allResourcesByModuleMap.put(module, mapper.selectProjectResources(module.getUuid()));
+    }
+
+    // and then proceed with the batch UPDATE at once
+    for (ResourceDto module : modules) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Combine this loop with the one that starts on line 137. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3047)